### ISSUE TITLE
Multiple sources on same domain

### ISSUE THREAD

**Issue by MyrikLD** [pypa/pipenv#4672]
### Issue description

ResolutionFailure on install packages from multiple sources with equal domains

### Expected result

Package installing

### Actual result

Locking [dev-packages] dependencies...
Building requirements...
Resolving dependencies...
✔ Success! 
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
✘ Locking Failed! 
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/resolver.py", line 741, in _main
[ResolutionFailure]:       resolve_packages(pre, clear, verbose, system, write, requirements_dir, packages, dev)
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/resolver.py", line 702, in resolve_packages
[ResolutionFailure]:       results, resolver = resolve(
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/resolver.py", line 684, in resolve
[ResolutionFailure]:       return resolve_deps(
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/utils.py", line 1395, in resolve_deps
[ResolutionFailure]:       results, hashes, markers_lookup, resolver, skipped = actually_resolve_deps(
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/utils.py", line 1108, in actually_resolve_deps
[ResolutionFailure]:       resolver.resolve()
[ResolutionFailure]:   File "/usr/lib/python3.9/site-packages/pipenv/utils.py", line 833, in resolve
[ResolutionFailure]:       raise ResolutionFailure(message=str(e))
[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
 Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
ERROR: Could not find a version that matches package_two (from -r /tmp/pipenvamergd9lrequirements/pipenv-djbch16n-constraints.txt (line 14))
No versions found
Were `https://pypi.org/simple` or `https://__token__:***@gitlab.com/api/v4/projects/123/packages/pypi/simple` or `https://__token__:***@gitlab.com/api/v4/projects/456/packages/pypi/simple` reachable?

### Steps to replicate

pipenv lock

-------------------------------------------------------------------------------

<details><summary>$ pipenv --support</summary>

Pipenv version: `'2020.11.15'`

Pipenv location: `'~/.local/share/virtualenvs/scraper-mWBbuJI4/lib/python3.9/site-packages/pipenv'`

Python location: `'~/.local/share/virtualenvs/scraper-mWBbuJI4/bin/python'`

Python installations found:

  - `3.9.2`: `~/.local/share/virtualenvs/scraper-mWBbuJI4/bin/python3`
  - `3.9.2`: `~/.local/share/virtualenvs/scraper-mWBbuJI4/bin/python3.9`
  - `3.9.2`: `~/.local/share/virtualenvs/scraper-mWBbuJI4/bin/python3`
  - `3.9.2`: `~/.local/share/virtualenvs/scraper-mWBbuJI4/bin/python3.9`
  - `3.9.2`: `/usr/bin/python3.9`
  - `3.9.2`: `/usr/bin/python3`
  - `3.8.8`: `/usr/bin/python3.8`
  - `2.7.18`: `/usr/bin/python2.7`
  - `2.7.18`: `/usr/bin/python2`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.9.2',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '5.11.10-1-MANJARO',
 'platform_system': 'Linux',
 'platform_version': '#1 SMP PREEMPT Thu Mar 25 17:14:55 UTC 2021',
 'python_full_version': '3.9.2',
 'python_version': '3.9',
 'sys_platform': 'linux'}
```

System environment variables:

  - `HOME`
  - `DBUS_SESSION_BUS_ADDRESS`
  - `XDG_CURRENT_DESKTOP`
  - `LC_MONETARY`
  - `MONGO_USERNAME`
  - `PATH`
  - `PIP_PYTHON_PATH`
  - `REDIS_PORT`
  - `SHELL`
  - `XDG_MENU_PREFIX`
  - `PIPENV_ACTIVE`
  - `LC_PAPER`
  - `SSH_AGENT_PID`
  - `QT_AUTO_SCREEN_SCALE_FACTOR`
  - `XDG_VTNR`
  - `USER`
  - `PIP_DISABLE_PIP_VERSION_CHECK`
  - `LC_TIME`
  - `XDG_SESSION_TYPE`
  - `LC_MEASUREMENT`
  - `XDG_SEAT`
  - `DESKTOP_SESSION`
  - `GTK_MODULES`
  - `LC_IDENTIFICATION`
  - `PYTHONDONTWRITEBYTECODE`
  - `MOTD_SHOWN`
  - `LC_NAME`
  - `LC_NUMERIC`
  - `TERM`
  - `GDK_BACKEND`
  - `XDG_SESSION_CLASS`
  - `GTK3_MODULES`
  - `XDG_CONFIG_DIRS`
  - `PWD`
  - `__pipenv_fish_initial_pwd`
  - `SESSION_MANAGER`
  - `XDG_SESSION_ID`
  - `EDITOR`
  - `LOGNAME`
  - `MONGO_PASSWORD`
  - `QT_QPA_PLATFORMTHEME`
  - `XDG_RUNTIME_DIR`
  - `REDIS_HOST`
  - `VTE_VERSION`
  - `LC_ADDRESS`
  - `LC_TELEPHONE`
  - `XAUTHORITY`
  - `MKLROOT`
  - `XDG_DATA_DIRS`
  - `JOURNAL_STREAM`
  - `DISPLAY`
  - `MAIL`
  - `LANG`
  - `SSH_AUTH_SOCK`
  - `SHLVL`
  - `COLORTERM`
  - `LS_COLORS`
  - `INVOCATION_ID`
  - `MONGO_HOST`
  - `GTK2_RC_FILES`
  - `VIRTUAL_ENV`
  - `PIP_SHIMS_BASE_MODULE`
  - `PYTHONFINDER_IGNORE_UNSUPPORTED`

Pipenv–specific environment variables:

 - `PIPENV_ACTIVE`: `1`

Debug–specific environment variables:

  - `PATH`: `/home/linuxbrew/.linuxbrew/opt/python@3.8/bin:package_two/.local/share/virtualenvs/scraper-mWBbuJI4/bin:package_two/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/var/lib/flatpak/exports/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl`
  - `SHELL`: `/bin/fish`
  - `EDITOR`: `/usr/bin/nano`
  - `LANG`: `en_US.UTF-8`
  - `PWD`: `/work/upsilon/liveart/scraper/laa_scrapy`
  - `VIRTUAL_ENV`: `package_two/.local/share/virtualenvs/scraper-mWBbuJI4`

---------------------------

Contents of `Pipfile` ('/work/upsilon/liveart/scraper/Pipfile'):

```toml
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[[source]]
url = "https://__token__:***@gitlab.com/api/v4/projects/123/packages/pypi/simple"
verify_ssl = true
name = "scraper"

[[source]]
url = "https://__token__:***@gitlab.com/api/v4/projects/456/packages/pypi/simple"
verify_ssl = true
name = "api"

[packages]
package_one = {version="*", index="api"}
package_two = {version="*", index="scraper"}

[requires]
python_version = "3"

[pipenv]
allow_prereleases = true

```

</details>

**Comment by matteius** [comment]
@MyrikLD Since this involves private Pypis can you re-test with the latest version `2022.1.8`?

**Comment by matteius** [comment]
I suspect the newest version of `2022.4.8` would not have this issue.

**Comment by ThomasSteinbach** [comment]
I am facing the same problem. Also with version `2022.4.8`. The last working version for me is `2021.11.23`.

**Comment by matteius** [comment]
@ThomasSteinbach what is your example Pipfile look like?

**Comment by ThomasSteinbach** [comment]
@matteius here you are. I only replaced some names due to confidentialy:

```
[[source]]
url = "https://repo.example.com/artifactory/api/pypi/default-pypi-3rdparty/simple"
verify_ssl = true
name = "default-pypi-3rdparty"

[[source]]
url = "https://repo.example.com/artifactory/api/pypi/myteam-cdk-pypi-dev/simple"
verify_ssl = true
name = "myteam-cdk-pypi-dev"

[[source]]
url = "https://repo.example.com/artifactory/api/pypi/myteam-cdk-pypi-prd/simple"
verify_ssl = true
name = "myteam-cdk-pypi-prd"

[[source]]
url = "https://repo.example.com/artifactory/api/pypi/myteam-gcip-pypi-dev/simple"
verify_ssl = true
name = "myteam-gcip-pypi-dev"

[[source]]
url = "https://repo.example.com/artifactory/api/pypi/myteam-gcip-pypi-prd/simple"
verify_ssl = true
name = "myteam-gcip-pypi-prd"

[packages]
# Minor versions in CDK might not be a stable change.
# That is why we only auto update micro version.
aws-cdk-lib = "~=2.18.0"
setuptools-git-versioning = "~=1.8"
boto3 = "~=1.20"
pyaml = ">=21"
toml = "~=0.10"
jinja2 = "~=3.0"

[dev-packages]
myteam-gcip = "~=2.0"
black = "*"
isort = "*"
flake8 = "*"
mypy = "*"
pytest-cov = "*"
pytest-mock = "*"
tomli = "~=2.0"

[requires]
python_version = "3.7"
```

**Comment by matteius** [comment]
@ThomasSteinbach Your problem is different from the original issue report.  You will need to pin your package to indexes whereby the first index in your Pipfile is the default that all unspecified packages will use.   This first would have been an issue for you in `2022.3.23` https://pipenv.pypa.io/en/latest/changelog/#id20 due to 

> Patched our vendored Pip to fix: Pipenv Lock (Or Install) Does Not Respect Index Specified For A Package. [#4637](https://github.com/pypa/pipenv/issues/4637)

For more context see updated documentation on indexes:  https://pipenv.pypa.io/en/latest/advanced/#specifying-package-indexes

See also:
- https://github.com/pypa/pipenv/issues/5021
- https://github.com/pypa/pipenv/issues/5022
- https://github.com/pypa/pipenv/issues/5023

**Comment by matteius** [comment]
@ThomasSteinbach So in your case update to use:

`myteam-gcip = {version= "~=2.0", index="myteam-gcip-pypi-prd"}`

**Comment by matteius** [comment]
@MyrikLD If you determine this is still an issue with the latest versions of pipenv and would like more support, just comment back with additional details.  Until then I am closing this issue.

**Comment by MyrikLD** [comment]
Thank you, fixed in `2022.4.8`
